src/execution_plan/ops/shared/update_functions.c

```diff
@@ -8,6 +8,26 @@
 #include "../../../errors.h"
 #include "../../../query_ctx.h"
 #include "../../../datatypes/map.h"
+#include "../../../datatypes/array.h"
+
+// recursively validate that an array property does not contain
+// a non-array, non-primitive type
+// returns true if property is valid
+static bool _ValidateArrayProperty(SIValue arr) {
+	ASSERT(SI_TYPE(arr) == T_ARRAY);
+
+	uint array_len = SIArray_Length(arr);
+	for(uint i = 0; i < array_len; i++) {
+		SIValue elem = SIArray_Get(arr, i);
+		// each element must be a primitive or an array
+		if(!(SI_TYPE(elem) & SI_VALID_PROPERTY_VALUE)) return false;
+
+		// recursively check nested arrays
+		if(SI_TYPE(elem) == T_ARRAY) return _ValidateArrayProperty(elem);
+	}
+
+	return true;
+}
```

Comment:
I think it would be cleaner if we introduce `bool containType(SIArray arr, SIType t)` to SIArray,
this way we can call this function using bitwise or to describe all forbidden types. also it seems like we might have missed cases where we set a property to an invalid value upon entity creation.